### DISCUSSION TITLE
Type of transacting prisma (e.g. tx)?

### DISCUSSION THREAD

**Discussion by olalonde** [prisma/prisma#20924] [Q&A]
### Question

In the [interactive transactions](https://www.prisma.io/docs/concepts/components/prisma-client/transactions#interactive-transactions) example, let's say I wanted to split up the transaction in separate functions. I would need to pass the `tx` object, but I'm not sure if its type is exported by Prisma. e.g. (see WhatTypeHere below):

```typescript
import { PrismaClient } from '@prisma/client'
const prisma = new PrismaClient()

async function incrementBalance(tx: WhatTypeHere, amount: number, to: string) {
    const recipient = await tx.account.update({
      data: {
        balance: {
          increment: amount,
        },
      },
      where: {
        email: to,
      },
    })
    return recipient
}

async function transfer(from: string, to: string, amount: number) {
  return await prisma.$transaction(async (tx) => {
    // 1. Decrement amount from the sender.
    const sender = await tx.account.update({
      data: {
        balance: {
          decrement: amount,
        },
      },
      where: {
        email: from,
      },
    })

    // 2. Verify that the sender's balance didn't go below zero.
    if (sender.balance < 0) {
      throw new Error(`${from} doesn't have enough to send ${amount}`)
    }

    // 3. Increment the recipient's balance by amount
    const recipient = await incrementBalance(tx, amount, to)

    return recipient
  })
}

async function main() {
  // This transfer is successful
  await transfer('alice@prisma.io', 'bob@prisma.io', 100)
  // This transfer fails because Alice doesn't have enough funds in her account
  await transfer('alice@prisma.io', 'bob@prisma.io', 100)
}

main()
```

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by olalonde** [comment]
Related: https://github.com/prisma/prisma/issues/10880

**Reply by ludralph** [reply]
Hi @olalonde 👋 

Apologies for the late reply here. I think the type you are looking for is `Prisma.TransactionClient`
```js
async function incrementBalance(tx: Prisma.TransactionClient, amount: number, to: string) {
    const recipient = await tx.account.update({
      data: {
        balance: {
          increment: amount,
        },
      },
      where: {
        email: to,
      },
    })
    return recipient
}
```

`WhatTypeHere` will be replaced with `Prisma.TransactionClient`

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by danbrakeley** [reply]
I ended up on this page from a search, and `Prisma.TransactionClient` worked for me.

**Reply by peretch** [reply]
I have a custom type for this cases, but it doesn't work if you use `$extends`

```
import { PrismaClient } from '@prisma/client';
import { DefaultArgs } from '@prisma/client/runtime/library';
import { Prisma } from 'prisma/prisma-client';

type OmitPrismaClient = Omit<
  PrismaClient<Prisma.PrismaClientOptions, never, DefaultArgs>,
  '$connect' | '$disconnect' | '$on' | '$transaction' | '$use' | '$extends'
>;

type PrismaTransaction = OmitPrismaClient;

export { PrismaTransaction };

```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!

**Comment by afonsomatos** [comment]
same
